pyccel/codegen/printing/ccode.py

```diff
@@ -1238,11 +1274,15 @@ def get_declare_type(self, expr):
         rank  = expr.rank
 
         if rank > 0:
-            if expr.is_ndarray or isinstance(expr.class_type, HomogeneousContainerType):
-                if expr.rank > 15:
+            if isinstance(class_type, NumpyNDArrayType):
+                if rank > 15:
                     errors.report(UNSUPPORTED_ARRAY_RANK, symbol=expr, severity='fatal')
+                    return dtype
                 self.add_import(c_imports['ndarrays'])
                 dtype = 't_ndarray'
+            elif isinstance(class_type, (HomogeneousSetType, HomogeneousListType)):
+                dtype = self.find_in_dtype_registry(expr.class_type)
+                return dtype
```

Comment:
Should the docstring be updated as well?

Reply:
Yes probably

Reply:
what do you think about this docstring:
```python
"""
        Find the corresponding C dtype in the dtype_registry or the STC container type.

        This function searches for the corresponding C data type in the `dtype_registry`.
        If the provided dtype is a container (like `HomogeneousSetType` or
        `HomogeneousListType`), it identifies the scalar type of the container and returns
        the appropriate type for the `STC` container.
        It raises a `PYCCEL_RESTRICTION_TODO` error if the dtype is not found in the registry.

        Parameters
        ----------
        dtype : DataType or ContainerType
            The data type of the expression. This can be a fixed-size numeric type,
            a primitive type, or a container type.

        Returns
        -------
        str
            The code which declares the data type in C or the corresponding `STC` container
            type.

        Raises
        ------
        PYCCEL_RESTRICTION_TODO
            If the dtype is not found in the dtype_registry.
        """
```

Reply:
you can check the updated version here e620bc4552e3560084f6115b353e6919a959d655

Reply:
Looks good. It seems like it should talk about returning the type but this is probably more accurate as it doesn't work for arrays :+1: